### DISCUSSION TITLE
Limit switches for X and Y axis

### DISCUSSION THREAD

**Discussion by mvackel** [Paciente8159/uCNC#490] [Config and issues]
I'm not being able to use the limit switches for axis X and Y. The Z switch is working but inverted and the pullup resistors seems to be disabled.

I used the files generated by the web config, with the defaults for the limits. When I try to use the Setup Wizard in UGS, I can't make the X and Y switches work. And the graphic display shows the corresponding alarms.

I've checked the pins/ports used in the config that are the following:
                   Port        ISR        Mega Pin        Ramps
Limit X         5B           -6            D3                 Xmin
Limit Y         1J              1            D14               Ymin
Limit Z         3D           -4            D18               Zmin

(Probe          2D            -3            D19               Zmax )
(I'm not using Probe now, but I will)

The Z switch works inverted (alarms when set to LOW). switches  X and Y do not work. If I invert the switches, X clears and Y stays triggered.
And all pullup resistors seems to be disabled.

How can I correct this?

**Comment by Paciente8159** [comment]
> I've checked the pins/ports used in the config that are the following:
> Port ISR Mega Pin Ramps
> Limit X 5B -6 D3 Xmin
> Limit Y 1J 1 D14 Ymin
> Limit Z 3D -4 D18 Zmin

This is not correct. Limit X is E5 pin. Not B5.

### About limit switches
This depends a lot on your type of switches you are using (NO or NC) and the ref that they connect to (VCC, GND or float).

So first you need to answer these 2 questions:

- When the switch **IS NOT** pressed what is the reference of the signal pin? (VCC, GND or float)
- When the switch **IS** pressed what is the reference of the signal pin? (VCC, GND or float)

There are multiple variations to this setting but my recommendation is to:

- use simple switches or opto-isolated ones using the COMMON (connected to GND) and NC pins
- enable pull ups or have either external pull ups

 Why connect this way you may ask?

By setting enabling the input pins pull up you will ensure that if you forget to wire your limits or they malfunction in someway your pin input will always see a logic HIGH value and this will increase the security of your system.
Using the NC pin and the COMMON connected to GND means that the limits will short the input to GND when it's not pressed. As soon as it is pressed it will open the connection leaving the pin in a float state. The internal pull up ensures that the unconnected pin reads a HIGH logic state and triggers the alarm.

But you may have other type of switches or configurations and you can adapt to them by:

- If you have external pull up or pull down switch circuit disable internal pull ups. Otherwise always leave them enabled to ensure the pin does not transition (due to environmental noise like EMF).
- You can invert the internal logic of the switches states by masking the limits with setting [$5](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#5----limit-pins-invert-mask)

**Comment by mvackel** [comment]
"This is not correct. Limit X is E5 pin. Not B5."

-- Ok. I made a mistake when typing.

-- About the X & Y limits, I tested the Mega board alone (took off the ramps board) and tried the following  (all with $5 = 0):
```
D3 (PE5) -> GND (jumper to GND)       ->  UGS Alarms:    XYZP
D3       -> VCC  (jumper to VCC)      -> XYZP
D14 (TX3) (PJ1) -> GND                -> XYZP
D14             -> VCC                -> XYZP

D18 (TX1) (PD3) -> GND               -> XY-P
D18             -> VCC               -> XYZP
D19 (RX1) (PD2) -> GND               -> XYZ-
D19             -> VCC               -> XYZP
```
It seems that only Z and Probe limit switches are working and X &  Y are disabled somehow....

**Reply by Paciente8159** [reply]
Maybe it's a configuration file bug out of the web builder.
Please send both override files and the json config in a zip so I can check them.

**Comment by mvackel** [comment]
Thanks. 

[boardmap_overrides.txt](https://github.com/Paciente8159/uCNC/files/12543179/boardmap_overrides.txt)
[cnc_hal_overrides.txt](https://github.com/Paciente8159/uCNC/files/12543180/cnc_hal_overrides.txt)
[ucnc_build.txt](https://github.com/Paciente8159/uCNC/files/12543186/ucnc_build.txt)

**Comment by Paciente8159** [comment]
Ok I now understand why X and Y are not changing for you.
You have both X and X2 and Y and Y2 limits defined and enabled.
You will always have to have them connected otherwise the limit alarm will always be triggered.

In your case you shorted X but not X2 too so the alarm was always active. The same for Y.

Disable X2 limit if you are not using it and the same for Y2.

You can confirm this if you `Enable extra settings store commands ($SS, $SL, $SR) and pin state print commands ($P)` option on the web config.

This will enable command $P that prints the current pin states.

**Comment by mvackel** [comment]
Hello,
This "almost" worked... You can't disable X2 because the display stops working.... :-(
Disabling Y2 and/or Z2 works normally, but not X2. Just commenting the line from the cnc_hal_overrides.h:
   #define LIMIT_X2_DISABLE
is enough to bring the display alive again.

**Reply by Paciente8159** [reply]
Again that is very odd. I don't have the board with me but tomorrow I'll be able to test and give you feedback.

**Reply by Paciente8159** [reply]
The problem seems to be on the cnc_hal_config.h file that causes the system to simply freeze.
When this happens the status led stops flashing (it's an indicator that the system stalled).
I'm investigating the issue.

**Reply by Paciente8159** [reply]
Ok. The issue is not web config related. It's actually the individual limits disable options that causes the code to simply stall (I'm still trying to understand why and if it's only on the AVR or if other architectures are affected or not. On the virtual emulator these options work just fine).

You can work around this by going to the Board tab in the web config and undefine both X2 and Y2 pins definitions instead of using the disable option on the HAL tab (i tested this and it works).

I'm going to hide that option on the web config for now until I figure this out.

Thanks.

**Comment by mvackel** [comment]
It works! Great. I was able to make the machine work with motors and limits.  Thanks!

(Now I have another problem with 4 axis - will open another thread)
